File: hIndex_countingSort.py
# Approach: Bucket sort or counting sort
# We can use bucket sort, and to do so -> we start with
# total buckets of length n+1 where n is len of citations array
# Each index will store count of how many papers have citations that many time
# for any number greater than n, we count it all in the last bucket.
# Then we will iterate over buckets array in reverse order adding each count to a sum and as soon as index and sum value at that index becomes equal or less, we pick that as our answer.
# TC: O(N)
# SC: O(N)
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        N = len(citations)
        buckets = [0] *  (N + 1)
        
        for _, c in enumerate(citations):
            if c > N:
                buckets[N] += 1
            else:
                buckets[c] += 1
        # print(buckets)
        ssum = 0
        for i in range(N, -1, -1):
            ssum += buckets[i]
            if ssum >= i:
                return i
        return -1

File: H-Index_countingSort.java
//counting sort
//Time Complexity:O(n)
//SPace Complexity:O(1)
//Approach- Counting sort is used to store the count of occurrance of each index in the citations array(as a hashmap). All those indices whose values are greater than or equal to the value of the citations array length are added to the last index of the count array. Then this temp array is iterated from behind and if I find any index whose count value is equal to or greater than the current index value, then that index would become the h index.

class Solution {
    public int hIndex(int[] citations) {
        int total=0;
        int n=citations.length;
        int[] temp=new int[n+1];
        for(int c:citations){
            if(c>=n){
                temp[n]++;
            }
            else{
                temp[c]++;
            }
        }
        for(int i=n;i>=0;i--){
            total+=temp[i];
            if(total>=i){
                return i;
            }
        }
        return 0;
    }
} 